From 582a488b4530af3d857ad5e4e9ad5eb22471dd6e Mon Sep 17 00:00:00 2001
From: Valentine Barshak <valentine.barshak@cogentembedded.com>
Date: Thu, 1 Aug 2019 23:35:33 +0300
Subject: [PATCH 04/19] clk: renesas: r8a77970-cpg-mssr: Fix SD clock

This uses custom divisor ratio table and mask value for V3M, because
the SD-IF0 register settings differ from all other R-Car Gen3 variants.

Signed-off-by: Valentine Barshak <valentine.barshak@cogentembedded.com>
---
 drivers/clk/renesas/r8a77970-cpg-mssr.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/clk/renesas/r8a77970-cpg-mssr.c b/drivers/clk/renesas/r8a77970-cpg-mssr.c
index 68ff20d..acaec8f 100644
--- a/drivers/clk/renesas/r8a77970-cpg-mssr.c
+++ b/drivers/clk/renesas/r8a77970-cpg-mssr.c
@@ -198,6 +198,11 @@ static const struct mstp_stop_table r8a77970_mstp_table[] = {
 	{ 0x00000000, 0x0, 0x00000000, 0 },
 };
 
+static const struct cpg_sd_div_table r8a77970_sd_div_table[] = {
+	{  4 << 4,  8 }, {  5 << 4, 12 }, {  6 << 4, 16 }, {  7 << 4, 18 },
+	{  8 << 4, 24 }, { 10 << 4, 36 }, { 11 << 4, 48 }, { 12 << 4, 10 },
+};
+
 static const void *r8a77970_get_pll_config(const u32 cpg_mode)
 {
 	return &cpg_pll_configs[CPG_PLL_CONFIG_INDEX(cpg_mode)];
@@ -210,6 +215,9 @@ static const struct cpg_mssr_info r8a77970_cpg_mssr_info = {
 	.mod_clk_size		= ARRAY_SIZE(r8a77970_mod_clks),
 	.mstp_table		= r8a77970_mstp_table,
 	.mstp_table_size	= ARRAY_SIZE(r8a77970_mstp_table),
+	.sd_div_table		= r8a77970_sd_div_table,
+	.sd_div_table_size	= ARRAY_SIZE(r8a77970_sd_div_table),
+	.sd_div_mask		= 0xff << 4,
 	.reset_node		= "renesas,r8a77970-rst",
 	.reset_modemr_offset	= CPG_RST_MODEMR,
 	.extalr_node		= "extalr",
-- 
2.7.4

